New and improved content management commands

jamalex opened the pull request:
- Some fixes to the transfer module
- Addition of `exportcontent` and `exportchannel` commands
- Rename of `retrievecontent` to `importcontent`

jamalex commented on kolibri/content/management/commands/importcontent.py:
```diff
@@ -56,47 +56,54 @@ def add_arguments(self, parser):
+                        filetransfer = transfer.FileDownload(url, dest)
+                    elif method == "local":
+                        srcpath = paths.get_content_storage_file_path(filename, datafolder=options["directory"])
+                        filetransfer = transfer.FileCopy(srcpath, dest)
```
Given how I implemented the transfer module, this part is the only difference between the `network` and `local` versions, so I just consolidated them.

jamalex replied:
Not sure I fully understand the suggestion -- these strings are coming in from the command line parser (`kolibri manage importcontent -- *network* <channel_id>` vs `kolibri manage importcontent -- *local* <channel_id> <datapath>`), so I have to conditionalize on a string.

Oh, I see -- I was thinking this was the part inside the `handle_async` method. But I pass those strings into here, instead of having two separate functions, one for local and one for network. The reason I couldn't cleanly separate the logic (and remain DRY) is that I don't want to run lines [86-91](https://github.com/jamalex/kolibri/blob/4715c99f3c4067c0425ad7d45ffde81e88df5f75/kolibri/content/management/commands/importcontent.py#L86) until the check for line [81](https://github.com/jamalex/kolibri/blob/4715c99f3c4067c0425ad7d45ffde81e88df5f75/kolibri/content/management/commands/importcontent.py#L81) has passed (since creating the `FileDownload` object already creates a connection to the server, which needs to happen before we can do things like `filetransfer.total_size` on line [95](https://github.com/jamalex/kolibri/blob/4715c99f3c4067c0425ad7d45ffde81e88df5f75/kolibri/content/management/commands/importcontent.py#L95)).

So... I couldn't see a way to separate them without duplicating a large portion of the code from this method.

jamalex replied:
But at this point, instead of string identifiers, I could use constants, or, if we don't imagine there being more than two methods, a boolean? And then `handle_async` would be the only one checking the method, and passing in that constant/boolean flag?

benjaoming replied:
I agree that it's fine to have it in one method to avoid code duplication, but you can still keep that method internal as in `_download` and only call public methods with a clean interface like `network_download()` ?

jamalex replied:
Ah, I see what you're suggesting now. So, even though they don't have the `_` prefix, I'd been thinking of all these methods as "internal", as I don't think it's common practice (outside tests, at least) to instantiate a `Command` object and call methods on it. I can add the `_` for explicitness, however.

aronasorman commented on kolibri/content/management/commands/importcontent.py:
```diff
@@ -56,47 +56,54 @@ def add_arguments(self, parser):
+
+                            for chunk in filetransfer:
                                 length = len(chunk)
                                 overall_progress_update(length)
```
Note that the update function takes an additional parameter (can be anything that's JSON serializable), which is then sent to the frontend.

aronasorman replied:
No need to change this, just something to remember in the future.

benjaoming commented on kolibri/content/management/commands/importcontent.py:
```diff
@@ -56,47 +56,54 @@ def add_arguments(self, parser):
+                        filetransfer = transfer.FileDownload(url, dest)
+                    elif method == "local":
+                        srcpath = paths.get_content_storage_file_path(filename, datafolder=options["directory"])
+                        filetransfer = transfer.FileCopy(srcpath, dest)
```
The string identifiers are somewhat a tiny bit ugly :) - maybe better to declare methods, like `network_download()` and have that call `self._download(method="network")` to internalize this.

benjaoming commented on kolibri/content/management/commands/importcontent.py:
```diff
@@ -56,47 +56,54 @@ def add_arguments(self, parser):
     def handle_async(self, *args, **options):
         if options['command'] == 'network':
-            self.handle_network_download(*args, **options)
+            self.handle_transfer("network", *args, **options)
```
a method prefix `handle_`? :) Methods do kind of handle stuff, but it's better to find method names that contains their primary purpose in the beginning, for the sake of readability, but not least auto-completion IDE stuff.

jamalex replied:
The `handle_async` method is @aronasorman's extra method on `AsyncCommand`, to parallel the `handle` method on `BaseCommand`. These `handle_` methods were similarly meant to indicate the top-level handlers for taking care of those command-line subcommands. But especially now that it's just one, it could just as easily be renamed to something like `transfer_content`.

benjaoming commented on kolibri/content/management/commands/exportcontent.py:
```diff
@@ -0,0 +1,50 @@
+                    dest = paths.get_content_storage_file_path(filename, datafolder=data_dir)
+
+                    # if the file already exists, add its size to our overall progress, and skip
+                    # TODO(jamalex): could do md5 checks here instead, to be ultra-safe
```
Idea....

It would be nice to take these `TODO`'s to the test, given how much unsolved stuff that ended up in the KA Lite codebase: Either create a Github issue immediately or implement the missing part or not do anything at. I mean, if it's supposed to be done, let's give it an issue.. or solve it immediately if the solution is easier than creating an issue :)

jamalex replied:
This one was less of a "MUST DO" and more of a "IF WE WANT TO DO MD5 CHECKING, IT WOULD GO HERE" note, for future reference. :)

Adding md5 checking here would be easy, but it's not necessarily a good idea, given the substantial performance penalty. I can just remove the TODO...

benjaoming replied:
Yup, I know we haven't discussed it, but I think I might be in favour of not using `TODOs` in the codebase at all, as a kind of principle..

benjaoming commented:
Great stuff! I was wondering how it ties into our previous thoughts about import/export commands? Maybe there's some terminology to keep an eye out for.. and maybe there's some modularity?

The ideas for import/export also encompassed user data.

See: https://github.com/learningequality/ka-lite/issues/4311

jamalex commented:
> I was wondering how it ties into our previous thoughts about import/export commands? Maybe there's some terminology to keep an eye out for.. and maybe there's some modularity?

The discussions we had in planning this stuff definitely took inspiration from our previous discussions about the import/export commands. That's also why I renamed `retrievecontent` to `importcontent` in this PR, for consistency -- we have `importchannel`, `exportchannel`, `importcontent`, and `exportcontent`.

In terms of modularity, would love to hear specifically what you're thinking. One piece of modularity added in a previous PR and further used here is [`kolibri.content.utils.transfer`](https://github.com/jamalex/kolibri/blob/4715c99f3c4067c0425ad7d45ffde81e88df5f75/kolibri/content/utils/transfer.py#L26), which encapsulates async/streaming file transfer logic (both downloads and copying) into reusable classes/methods, so that we can add logic for error handling etc in a single place and take advantage everywhere.

I also previously added the [`kolibri.core.discovery`](https://github.com/jamalex/kolibri/tree/4715c99f3c4067c0425ad7d45ffde81e88df5f75/kolibri/core/discovery) app, which encapsulates the logic for *discovering* places to import/export to and from. Right now it just does filesystem stuff, but that's where the peer discovery stuff will go too.

> The ideas for import/export also encompassed user data.

Yep! Those pieces aren't written yet, and won't go under the content app, but will use similar terminology, and leverage the `discovery` app for identifying sources/destinations.

We can then have some higher-level import/export commands which call the various app-specific 
commands.

We added the `KOLIBRI_DATA` folder name, which is a standard name for an external data folder, e.g. on a USB key, and the structure of that folder directly mirrors the `.kolibri` (KOLIBRI_HOME) structure (/content/storage/, /content/databases/, etc).

benjaoming commented:
Thanks for explaining @jamalex 

I don't really have any specific ideas to table - but it sounds really great that async transfers have been modularized, we can definitely use this for other purposes.

One simple idea would be if kolibri in the end would just have one "import", and one "export" command, and one produced a file that the other could import -- and that file could contain content, database, userdata, perhaps even plugins and configuration stuff. Kind of like Firefox sync or something. But the main purpose would probably be smaller deployments where - say - you want to put Kolibri on 10 different servers, and you don't have Ansible.... or Norton Ghost :)

jamalex commented:
> One simple idea would be if kolibri in the end would just have one "import", and one "export" command, and one produced a file that the other could import

Yeah, that's exactly the plan. Except rather than "a file", it would produce a directory. Exporting any of these things (individually) puts them into the appropriate place under the target data directory, and similarly for importing. We would version the schema of that data directory -- it would also mirror the structure of the `.kolibri` directory itself.

jamalex commented:
All suggestions made above have been addressed, along with two other important fixes:

- Completing the circle by [updating](https://github.com/learningequality/kolibri/pull/462/commits/365e06e98bcaa0b1d664b0c39252dd5605c8ab09) `importchannel` to match `importcontent`, and allow local channel DB import.
- [Fixing](https://github.com/learningequality/kolibri/pull/462/commits/b5afc856a48daac1c0a7f3b5c7e473023dc79c21) a critical bug in the `FileCopy` module that was leading to each block of data being written twice during file-to-file copies.

jamalex commented:
> Completing the circle by updating importchannel to match importcontent, and allow local channel DB import.

@indirectlylit @ralphiee22 @aronasorman: Note that this means the interface to the `importchannel` command has changed slightly, as it is now more similar to `importcontent`, allowing two different import modes (method can be either "network" or "local").
